File: src/cpp_actions/src/fibonacci_client.cpp
#include <functional>
#include <future>
#include <memory>
#include <string>
#include <sstream>

#include "my_ros2_interface/action/fibonacci.hpp"

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "rclcpp_components/register_node_macro.hpp"

class FibonacciActionClient : public rclcpp::Node
{
public:
  using Fibonacci = my_ros2_interface::action::Fibonacci;
  using GoalHandleFibonacci = rclcpp_action::ClientGoalHandle<Fibonacci>;
  

  explicit FibonacciActionClient(const rclcpp::NodeOptions & node_options = rclcpp::NodeOptions())
  : Node("fibonacci_action_client", node_options), goal_done_(false)
  {
    this->client_ptr_ = rclcpp_action::create_client<Fibonacci>(
      this,
      "fibonacci");
    
     //declare order parameter
    this->declare_parameter<int>("order", 2);

    //get order parameter
    this->get_parameter("order", order_);

    RCLCPP_INFO(this->get_logger(), "Action client node has been started");

    this->timer_ = this->create_wall_timer(
      std::chrono::milliseconds(500),
      std::bind(&FibonacciActionClient::send_goal, this));
  }

  bool is_goal_done() const
  {
    return this->goal_done_;
  }

  void send_goal()
  {
    using namespace std::placeholders;

    this->timer_->cancel();
    this->goal_done_ = false;

    if (!this->client_ptr_) {
      RCLCPP_ERROR(this->get_logger(), "Action client not initialized");
    }

    if (!this->client_ptr_->wait_for_action_server(std::chrono::seconds(10))) {
      RCLCPP_ERROR(this->get_logger(), "Action server not available after waiting");
      this->goal_done_ = true;
      return;
    }

    auto goal_msg = Fibonacci::Goal();
    goal_msg.order = order_;

    RCLCPP_INFO(this->get_logger(), "Sending goal %d", order_);

    auto send_goal_options = rclcpp_action::Client<Fibonacci>::SendGoalOptions();
    send_goal_options.goal_response_callback =
      std::bind(&FibonacciActionClient::goal_response_callback, this, _1);
    send_goal_options.feedback_callback =
      std::bind(&FibonacciActionClient::feedback_callback, this, _1, _2);
    send_goal_options.result_callback =
      std::bind(&FibonacciActionClient::result_callback, this, _1);
    this->client_ptr_->async_send_goal(goal_msg, send_goal_options);
  }

private:
  rclcpp_action::Client<Fibonacci>::SharedPtr client_ptr_;
  rclcpp::TimerBase::SharedPtr timer_;
  bool goal_done_ = true;
  int order_;

  void goal_response_callback(const GoalHandleFibonacci::SharedPtr & goal_handle)
  {
    if (!goal_handle) {
      RCLCPP_ERROR(this->get_logger(), "Goal was rejected by server");
    } else {
      RCLCPP_INFO(this->get_logger(), "Goal accepted by server, waiting for result");
    }
  }

  void feedback_callback(
    GoalHandleFibonacci::SharedPtr,
    const std::shared_ptr<const Fibonacci::Feedback> feedback)
  {
    std::stringstream ss;
    ss << "Next number in sequence received: ";
    for (auto number : feedback->partial_sequence) {
      ss << number << " ";
    }
    RCLCPP_INFO(this->get_logger(), ss.str().c_str());
  }

  void result_callback(const GoalHandleFibonacci::WrappedResult & result)
  {
    this->goal_done_ = true;
    switch (result.code) {
      case rclcpp_action::ResultCode::SUCCEEDED:
        break;
      case rclcpp_action::ResultCode::ABORTED:
        RCLCPP_ERROR(this->get_logger(), "Goal was aborted");
        return;
      case rclcpp_action::ResultCode::CANCELED:
        RCLCPP_ERROR(this->get_logger(), "Goal was canceled");
        return;
      default:
        RCLCPP_ERROR(this->get_logger(), "Unknown result code");
        return;
    }
    std::stringstream ss;
    ss << "Result received: ";
    for (auto number : result.result->result_sequence) {
      ss << number << " ";
    }
    RCLCPP_INFO(this->get_logger(), ss.str().c_str());
    rclcpp::shutdown();
  }
};  // class FibonacciActionClient

int main(int argc, char* argv[]){
    rclcpp::init(argc, argv);

    auto node = std::make_shared<FibonacciActionClient>();

    while (!node->is_goal_done()) {
    rclcpp::spin_some(node);
  }

    rclcpp::shutdown();
    return 0;
}

File: src/py_actions/py_actions/fibonacci_client.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from ros2_fndm_interface.action import Fibonacci
from rclpy.action import ActionClient
from rclpy.action.client import ClientGoalHandle

import time

class FibonacciClientNode(Node):
    def __init__(self):
        super().__init__('fibonacci_client_node')

        self.action_client_ = ActionClient(self, Fibonacci, 'fibonacci')

        # get goal parameter value
        self.declare_parameter('order', 2) # default 2
        self.order = self.get_parameter('order').value

        self.get_logger().info('Fibonacci action client has been started')

  
    def send_goal(self, order):
        """the action client sends the goal"""

        # wait for the server
        timeout_ = 1.0
        self.action_client_.wait_for_server(timeout_sec=timeout_)

        # create a goal
        goal = Fibonacci.Goal()
        goal.order = order

        # send the goal
        self.get_logger().info(f'sending the goal: {order}')

        # send_goal_async with feedback_callback. It returns future. when future is returned, we can add a callback
        self.action_client_.send_goal_async(goal, feedback_callback=self.goal_feedback_callback).add_done_callback(self.goal_response_callback)

    
    def goal_feedback_callback(self, feedback_msg):
        """gets feedback message from feedback_callback"""

        feedback = feedback_msg.feedback
        self.get_logger().info(f'Received feedback: {feedback.partial_sequence}')


    def goal_response_callback(self, future):
        """checks if the goal request was accepted"""

        self.goal_handle_ : ClientGoalHandle = future.result()

        # checks if the goal has been accepted
        if not self.goal_handle_.accepted:
            self.get_logger().info('Goal rejected ')
            return 
        
        # if the goal has been accepted we can request for the result
        self.get_logger().info('Goal accepted')

        self.goal_handle_.get_result_async().add_done_callback(self.goal_result_callback) 

    
    def goal_result_callback(self, future):
        """gets the result """
        result = future.result().result
        fibonacci_sequence = result.result_sequence

        self.get_logger().info(f'result: {fibonacci_sequence}')



def main(args=None):
    rclpy.init(args=args)
    node = FibonacciClientNode()
    
    # send goal
    node.send_goal(node.order)

    rclpy.spin(node)

    try:
        rclpy.shutdown()
    except:
        pass

if __name__=='__main__':
    main()



     